### PR TITLE
Fix memory leak in mbedtls_x509_csr_parse

### PR DESCRIPTION
**Pull request by catenacyber**
## Description
Fix memory leak in mbedtls_x509_csr_parse

## Status
READY

## Requires Backporting
Yes

Which branch? I did not check

## Migrations
No

## Additional comments
Found using oss-fuzz

## Todos
- [X] Tests
- [X] Documentation
- [X] Changelog updated
- [X] Backported

## Steps to test or reproduce
Allocated memory in previous call to mbedtls_pem_read_buffer does not get freed without this patch

### CODE REVIEW

**Comment by catenacyber** [comment]
CLA has just been submitted through mail

**Comment by RonEld** [comment]
@catenacyber Thank you for your contribution! As sent in mail, the CLA is being proccessed

**Comment by RonEld** [comment]
@catenacyber I have just opened on your name bug #1623 for tracking purposes

**Comment by irwir** [comment]
Could this be written in a shorter way?

```
#if defined(MBEDTLS_PEM_PARSE_C)
    mbedtls_pem_init( &pem );

    /* Avoid calling mbedtls_pem_read_buffer() on non-null-terminated string */
    if( buf[buflen - 1] == '\0' )
    {
        ret = mbedtls_pem_read_buffer( &pem,
                               "-----BEGIN CERTIFICATE REQUEST-----",
                               "-----END CERTIFICATE REQUEST-----",
                               buf, NULL, 0, &use_len );

        if( ret == 0 )
            /*
             * Was PEM encoded, parse the result
             */
            ret = mbedtls_x509_csr_parse_der( csr, pem.buf, pem.buflen );
        mbedtls_pem_free( &pem );
        if( ret != MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT ) //not sure if this line is necessary
            return( ret );
    }
#endif /* MBEDTLS_PEM_PARSE_C */

```

**Comment by sbutcher-arm** [comment]
Needs backporting to `mbedtls-2.7` and `mbedtls-2.1`.

**Comment by irwir** [comment]
@andresag01
> the value of ret after line 290 satisfies the following condition ret != 0 && ret == MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT

Function call in the line 287 could return `MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT`, therefore memory leak is possible. The issue existed since polarssl days.

Refactored code in [this message](https://github.com/ARMmbed/mbedtls/pull/1621#issuecomment-393823993) should never leak memory. If necessary, a new PR could be created.

**Comment by catenacyber** [comment]
@andresag01 
There is no such leak (for now) :

> pem context is initialised in lines 281

There is no allocation, just a memset
The allocation take place in `mbedtls_pem_read_buffer` but only after we are sure we will not return `MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT`

**Comment by catenacyber** [comment]
The code proposed by @irwir looks good to me.

This pull request was the minimum change to fix the leak.
I let you decide if you want refactoring in the same or another pull request

**Comment by sbutcher-arm** [comment]
Hi @catenacyber,

I'd like to go with @andresag01's suggestion, but I'm conscious that this is your PR. You can either fix it yourself, and resubmit, or if you're having more fun finding the problems than fixing them, @andresag01 can do the re-work. Either way, we will credit you for the work you did here!

Please let us know.

**Comment by catenacyber** [comment]
Here you go with one more commit to simplify the code.
I let you review it

File: library/x509_csr.c [hunk 1]
```diff
@@ -278,34 +278,24 @@ int mbedtls_x509_csr_parse( mbedtls_x509_csr *csr, const unsigned char *buf, siz
         return( MBEDTLS_ERR_X509_BAD_INPUT_DATA );
 
 #if defined(MBEDTLS_PEM_PARSE_C)
-    mbedtls_pem_init( &pem );
-
     /* Avoid calling mbedtls_pem_read_buffer() on non-null-terminated string */
-    if( buf[buflen - 1] != '\0' )
-        ret = MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT;
-    else
+    if( buf[buflen - 1] == '\0' ) {
```
Comment (andresag01) [inline library/x509_csr.c]: Please note that the Mbed TLS coding style is to place the `{` on the next line.

**Comment by sbutcher-arm** [comment]
I've backported this to `mbedtls-2.7` with #1772, and `mbedtls-2.1` with #1174.

Once those are approved, this can move to 'ready for merge'.

**Pull request merged**